build_tools/torchscript_e2e_heavydep_tests/aot_backward.py

```diff
@@ -0,0 +1,129 @@
+)
+from torch_mlir.passmanager import PassManager
+from torch_mlir.dialects.torch.importer.jit_ir import ClassAnnotator, ModuleBuilder
+
+
+class NeuralNet(nn.Module):
+    def __init__(self):
+        super(NeuralNet, self).__init__()
+        self.l1 = nn.Linear(10, 16) 
+        self.relu = nn.ReLU()
+        self.l2 = nn.Linear(16, 2)
+
+    def forward(self, x):
+        out = self.l1(x)
+        out = self.relu(out)
+        out = self.l2(out)
+        return out
+
+model = NeuralNet()
+
+criterion = nn.CrossEntropyLoss()
+optimizer = torch.optim.SGD(model.parameters(), lr=0.01)
+
+
+input = torch.randn(1, 10)
+labels = torch.randn(1, 2)
+
+def bench_model(mod):
+    iters = 1
+    for _ in range(iters):
+        optimizer.zero_grad()
+        output = mod(input)
+        loss = criterion(output, labels)
+        loss.backward()
+        optimizer.step()
+
+
+def save_mlir(module, name):
+    with open(name, "w") as text_file:
+        text_file.write(str(module))
+
+
+def generate_mlir_module(ts_script_module, annotations, name):
+    mb = ModuleBuilder()
+    class_annotator = ClassAnnotator()
+    class_annotator.exportNone(ts_script_module._c._type())
+    class_annotator.exportPath(ts_script_module._c._type(), ["forward"])
+    class_annotator.annotateArgs(
+        ts_script_module._c._type(), ["forward"], annotations
+    )
+    mb.import_module(ts_script_module._c, class_annotator)
+    mlir_module = mb.module
+    save_mlir(mlir_module, name)
+    mlir_module.dump()
+
+def change_fx_graph_return_to_tuple(fx_g: fx.GraphModule) -> fx.GraphModule:
+    for node in fx_g.graph.nodes:
+        if node.op == 'output':
+            # output nodes always have one argument
+            node_arg = node.args[0]
+            if isinstance(node_arg, list):
+                node.args = (tuple(node_arg),)
+    fx_g.graph.lint()
+    fx_g.recompile()
+    return fx_g
+
+def generate_annotation(inputs):
+    annotations_list = [None]
+    for i in inputs:
+        temp_list = []
+        temp_list.append(list(i.shape))
+        temp_list.append(i.dtype)
+        temp_list.append(True)
+        annotations_list.append(tuple(temp_list))
+    return annotations_list
+
+
+def ts_compiler_forward(fx_g: fx.GraphModule, inps):
+    fx_g = change_fx_graph_return_to_tuple(fx_g)
+    f = torch.jit.script(fx_g)
+    f = torch.jit.freeze(f.eval())
+    torch.jit.save(f, "forw.pt")
+    f = torch.jit.load("forw.pt")
+    print("The forward torch-mlir graph is:")
+    generate_mlir_module(
+        f, generate_annotation(inps), "forward.mlir"
+    )
+    return f
+
+
+def ts_compiler_backward(fx_g: fx.GraphModule, inps):
+    fx_g = change_fx_graph_return_to_tuple(fx_g)
+    f = torch.jit.script(fx_g)
+    f = torch.jit.freeze(f.eval())
+    torch.jit.save(f, "back.pt")
+    f = torch.jit.load("back.pt")
+    print("The backward torch-mlir graph is:")
+    generate_mlir_module(
+        f, generate_annotation(inps), "backward.mlir"
+    )
+    return f
+
+
```

Comment:
can you please check this in as an actual e2e test?

See the structure of https://github.com/llvm/torch-mlir/blob/main/build_tools/torchscript_e2e_heavydep_tests/minilm_sequence_classification.py

Notice:
```
@register_test_case(module_factory=lambda: getTracedRecursiveScriptModule())
def MiniLMSequenceClassification_basic(module, tu: TestUtils):
    module.forward(test_input)
```

Can you also please verify that you are able to run the resulting test (even if it fails) following the instructions at https://github.com/llvm/torch-mlir/tree/main/build_tools/torchscript_e2e_heavydep_tests/README.md

Reply:
This test just demonstrates the generation of the backward graph via AOT autograd, the intention to add it as an e2e test is not there.

Reply:
I consider it a requirement to actually add it as an e2e test. Then we can all work together towards running it which is a major project goal.